From ad08878c30cbab57b6ea8fa8f27572cc689dc2e8 Mon Sep 17 00:00:00 2001
From: Li Huafei <lihuafei1@huawei.com>
Date: Mon, 19 Jun 2023 20:01:33 +0800
Subject: [PATCH 10/36] [Huawei] x86/kprobes: Fix kprobe missing some debugging
 exceptions

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188915

--------------------------------

When fixing the kprobe debugging exception handling logic earlier, we
incorrectly assumed that if a debugging exception was kprobe triggered,
then regs->ip should be located in kprobe's instruction slot. If the
staked instruction was call or jump, then regs->ip would be the
destination address of the jump and not in the slot.

In fact, the correct and more direct way to determine this would be
based on the current state of the kprobe. kprobe will change the status
from KPROBE_HIT_ACTIVE to KPROBE_HIT_SS or KPROBE_REENTER before
single-step execution, so if the current status is KPROBE_HIT_ACTIVE,
its not a debug exception triggered by kprobe.

Fixes: c209afe67ca3 ("[Huawei] x86/kprobes: Fix the error judgment for debug exceptions")
Signed-off-by: Li Huafei <lihuafei1@huawei.com>
---
 arch/x86/kernel/kprobes/core.c | 9 +++++----
 1 file changed, 5 insertions(+), 4 deletions(-)

diff --git a/arch/x86/kernel/kprobes/core.c b/arch/x86/kernel/kprobes/core.c
index fd8d7d128807..352cf0a264d6 100644
--- a/arch/x86/kernel/kprobes/core.c
+++ b/arch/x86/kernel/kprobes/core.c
@@ -903,11 +903,12 @@ int kprobe_debug_handler(struct pt_regs *regs)
 	if (!cur || !cur->ainsn.insn)
 		return 0;
 
-	/* regs->ip should be the address of next instruction to
-	 * cur->ainsn.insn.
+	/* kprobe will change the status from KPROBE_HIT_ACTIVE to
+	 * KPROBE_HIT_SS or KPROBE_REENTER before single-step execution, so
+	 * if the current status is KPROBE_HIT_ACTIVE, its not a debug
+	 * exception triggered by kprobe.
 	 */
-	if (regs->ip < (unsigned long)cur->ainsn.insn ||
-	    regs->ip - (unsigned long)cur->ainsn.insn > MAX_INSN_SIZE)
+	if (kcb->kprobe_status == KPROBE_HIT_ACTIVE)
 		return 0;
 
 	resume_execution(cur, regs, kcb);
-- 
2.17.1

